### PR TITLE
libcephsqlite: shutdown RADOS in destructor

### PR DESCRIPTION
**Pull request by batrick**
Fixes: https://tracker.ceph.com/issues/50503
Fixes: https://tracker.ceph.com/issues/51372
Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by batrick** [comment]
before: https://pulpito.ceph.com/pdonnell-2021-06-25_19:04:40-rados:basic-master-distro-basic-smithi/
after: https://pulpito.ceph.com/pdonnell-2021-06-25_20:06:40-rados:basic-wip-pdonnell-testing-20210625.184028-distro-basic-smithi/

**Comment by neha-ojha** [comment]
> before: https://pulpito.ceph.com/pdonnell-2021-06-25_19:04:40-rados:basic-master-distro-basic-smithi/
> after: https://pulpito.ceph.com/pdonnell-2021-06-25_20:06:40-rados:basic-wip-pdonnell-testing-20210625.184028-distro-basic-smithi/

after has some failures too :(

**Comment by batrick** [comment]
Ya... not yet sure why.

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-06-27_03:50:15-rados:basic-wip-pdonnell-testing-20210627.023003-distro-basic-smithi/

**Comment by batrick** [comment]
Pretty sure this will be fixed by the latest commits. Couldn't reproduce the crash anymore testing locally for 8 hours. QA results should be in soon but sepia appears down.

**Comment by batrick** [comment]
> https://pulpito.ceph.com/pdonnell-2021-06-27_03:50:15-rados:basic-wip-pdonnell-testing-20210627.023003-distro-basic-smithi/

tests pass!

**Review by neha-ojha — APPROVED** [review]

File: src/libcephsqlite.cc [hunk 1]
```diff
@@ -807,8 +819,15 @@ LIBCEPHSQLITE_API int sqlite3_cephsqlite_init(sqlite3* db, char** err, const sql
...
-    appd->cct = nullptr;
-    sqlite3_vfs_register(vfs, 0);
+    if (int rc = sqlite3_vfs_register(vfs, 0); rc) {
+      delete appd;
+      free(vfs);
+      return rc;
+    }
+  }
+
+  if (int rc = std::atexit(cephsqlite_atexit); rc) {
```
Comment (tchaikov) [inline src/libcephsqlite.cc:829]: probably a better way to clean up app data is to hook a function to `xDlClose` in `sqlite3_cephsqlite_init()`?

Comment (batrick) [reply src/libcephsqlite.cc:829]: sqlite does not do any cleanup on error, it just calls `exit(3)`.

Comment (batrick) [reply src/libcephsqlite.cc:829]: `xDlClose` would be called in the UNIX VFS, in any case, as that's the VFS which loaded `libcephsqlite`.

**Review by tchaikov — APPROVED** [review]

**Pull request merged**